costmodel/promparsers.go

```diff
@@ -0,0 +1,350 @@
+package costmodel
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+
+	costAnalyzerCloud "github.com/kubecost/cost-model/cloud"
+	"k8s.io/klog"
+)
+
+// PromQueryResult contains a single result from a prometheus query
+type PromQueryResult struct {
+	Metric map[string]interface{}
+	Values []*Vector
+}
+
+func (pqr *PromQueryResult) GetString(field string) (string, error) {
+	f, ok := pqr.Metric[field]
+	if !ok {
+		return "", fmt.Errorf("%s field does not exist in data result vector", field)
+	}
+
+	strField, ok := f.(string)
+	if !ok {
+		return "", fmt.Errorf("%s field is improperly formatted", field)
+	}
+
+	return strField, nil
+}
+
+func (pqr *PromQueryResult) GetLabels() map[string]string {
+	result := make(map[string]string)
+
+	// Find All keys with prefix label_, remove prefix, add to labels
+	for k, v := range pqr.Metric {
+		if !strings.HasPrefix(k, "label_") {
+			continue
+		}
+
+		label := k[6:]
+		value, ok := v.(string)
+		if !ok {
+			klog.V(3).Infof("Failed to parse label value for label: %s", label)
+			continue
+		}
+
+		result[label] = value
+	}
+
+	return result
+}
+
+// NewQueryResults accepts the raw prometheus query result and returns an array of
+// PromQueryResult objects
+func NewQueryResults(queryResult interface{}) ([]*PromQueryResult, error) {
```

Comment:
@AjayTripathy This is my proposal for query results. It handles both `value` and `values` based on metrics, does careful checking for each property and propagates errors correctly. Helper function for retrieving labels from the metrics as well as getting specific string values. The query parsing in this file uses this, and I switched Network parsing over to this as well. It seems to reduce the noise of the parsing methods a good bit.